feat:[CDS-99739]: Sequential and parallel deployments (#7613)
harness/developer-hub: docs/continuous-delivery/x-platform-cd-features/sequential-and-parallel-deployments.md

---
title: Sequential and Parallel Deployments
description: How you can do sequential and parallel deployments with multi-service, multi-environment and multi-infrastructure
sidebar_position: 300
---

> Pipelines can be executed with multiple Services and Environments. These Services and Environments are combined into a Service & Environment (infrastructure) in a stage for execution.

Reviewer: ```suggestion
Pipelines can be executed with multiple services and environments. These services and environments are combined into a service & environment (infrastructure) in a stage for execution.
```

### Execution Sequence

Users can configure the sequence of deployment as Sequential or Parallel.

> When you click on **Deploy multiple Services** you see an option for **Deploy Services in Parallel?**. When you check that it will allow you to deploy your services parallelly.

Reviewer: ```suggestion
When you click on **Deploy multiple Services** you see an option for **Deploy Services in Parallel?**. When you check that it will allow you to deploy your services in parallel.
```

![](./static/service_parallel.png)

:::important note

> By default, the execution sequence is the parallel deployment for Services.

Reviewer: ```suggestion
By default, the execution sequence is parallel deployment for services.
```

:::

> When you click on **Deploy to multiple Environments or Infrastructures** you see an option for **Deploy to Environments or Infrastructures in parallel?**. When you check that it will allow you to deploy your infrastructure and environment parallelly.

Reviewer: ```suggestion
When you click on **Deploy to multiple Environments or Infrastructures** you see an option for **Deploy to Environments or Infrastructures in parallel?**. When you check that it will allow you to deploy your infrastructure and environment in parallel.
```

![](./static/environment_infrastructure.png)

> ### Expected Scenario

Reviewer: ```suggestion
### Expected Scenarios
```

> 1. Sequential Deployment in Multi Services and Sequential Multi-Infrastructure

Reviewer: ```suggestion
#### Sequential Deployment in Multi Services and Sequential Multi-Infrastructure
```

For example:

- **Services**: S1, S2
- **Environment**: E1
- **Infrastructure**: I1, I2

> Each service is sequentially deployed in collection of Infrastructure ordered in the YAML.

Reviewer: Should this be
```suggestion
Each service is sequentially deployed in a collection of Infrastructure instructed by the YAML.
```

Each service is deployed in one infrastructure first and then followed by the second infrastructure.

![](./static/service_infra_sequence.png)

The deployment sequence is as follows:
**`S1 I1, S2 I1, S1 I2, S2 I2`**

2. Parallel Deployment in Multi-Services and Sequential Multi-Infrastructure

For example:

- **Services**: S1, S2
- **Environment**: E1
- **Infrastructure**: I1, I2

Each service will be deployed parallelly in different infrastructure.

![](./static/service_parallel_infra_seq.png)

The deployment sequence is as follows:

> Here, 1 and 2 will be deployed parallelly and 3 and 4 will be deployed parallelly i.e same services are deployed parallelly in the different infrastructure. 1 and 3 will deployed sequentially and 2 and 4 will be deployed sequentially i.e different service in same infrastructure will deployed sequentially.

Reviewer: Please use in parallel over parallely. 
```suggestion
Here, 1 and 2 will be deployed in parallel and 3 and 4 will be deployed in parallel i.e the same services are deployed in parallel in each infrastructure. 1 and 3 will deployed sequentially and 2 and 4 will be deployed sequentially i.e different service in same infrastructure will deployed sequentially.
```

Reviewer: Is the first sentence here about services and the second about environments?

![](./static/Parallel_service_seq_infra.png)

3. Sequential Deployment in Multi-Services and Parallel Multi-Infrastructure

For example:

- **Services**: S1, S2
- **Environment**: E1
- **Infrastructure**: I1, I2

![](./static/service_parallel_infra_seq.png)

Each service is deployed sequentially, but the infrastructures are deployed in parallel.

The deployment sequence is as follows:

Here, 1 and 2 will be deployed parallelly and 3 and 4 will be deployed parallelly i.e different services are deployed parallelly in the same infrastructure. 1 and 3 will deployed sequentially and 2 and 4 will be deployed sequentially i.e same service will be deployed in different infrastructure sequentially.

![](./static/seq_service_parallel_infra.png)

4. Multi Service in Parallel and Multi Infrastructure in Parallel

- **Services**: S1, S2
- **Environment**: E1
- **Infrastructure**: I1, I2

Both services and infrastructure are deployed in parallel.

The deployment sequence is as follows:
**`S1 I1, S2 I1, S1 I2, S2 I2`**

> ![](./static/service_parallel_infra_parallel.png)

Reviewer: Is this the right image? it looks like its depicting a sequential deployment.

---
title: Sequential and Parallel Deployments
description: How you can do sequential and parallel deployments with multi-service, multi-environment and multi-infrastructure
sidebar_position: 300
---

Pipelines can be executed with multiple services and environments. These services and environments are combined into a service & environment (infrastructure) in a stage for execution.

## Execution Sequence

Users can configure the sequence of deployment as Sequential or Parallel.

> When you click on **Deploy multiple Services** you see an option for **Deploy Services in Parallel?**. When you check that it will allow you to deploy your services in parallel.

Reviewer 2: Remove the question mark in this line

![](./static/service_parallel.png)

:::important note
By default, the execution sequence is parallel deployment for services.
:::

> When you click on **Deploy to multiple Environments or Infrastructures** you see an option for **Deploy to Environments or Infrastructures in parallel?**. When you check that it will allow you to deploy your infrastructure and environment in parallel.

Reviewer 2: This line seems repeating the same words

![](./static/environment_infrastructure.png)

## Expected Scenarios

> ### Sequential Deployment in Multi Services and Sequential Multi-Infrastructure

Reviewer 2: Sequential Deployment in Multi Services and Multi-Infrastructure

> For example:

Reviewer 2: Rename this as Sample configuration

- **Services**: S1, S2
- **Environment**: E1
- **Infrastructure**: I1, I2

Each service is sequentially deployed in a collection of Infrastructure instructed by the YAML.

Each service is deployed in one infrastructure first and then followed by the second infrastructure.

![](./static/service_infra_sequence.png)

The deployment sequence is as follows:
**`S1 I1, S2 I1, S1 I2, S2 I2`**

> ### Parallel Deployment in Multi-Services and Sequential Multi-Infrastructure

Reviewer 2: Parallel Deployment in Multi-Services and Sequential in Multi-Infrastructure

For example:

- **Services**: S1, S2
- **Environment**: E1
- **Infrastructure**: I1, I2

Each service will be deployed parallelly in different infrastructure.

![](./static/service_parallel_infra_seq.png)

The deployment sequence is as follows:

> Here, 1 and 2 will be deployed in parallel and 3 and 4 will be deployed in parallel i.e the same services are deployed in parallel in each infrastructure. 1 and 3 will deployed sequentially and 2 and 4 will be deployed sequentially i.e different service in same infrastructure will deployed sequentially.

Reviewer 2: This is confusing

![](./static/Parallel_service_seq_infra.png)

### Sequential Deployment in Multi-Services and Parallel Multi-Infrastructure

For example:

- **Services**: S1, S2
- **Environment**: E1
- **Infrastructure**: I1, I2

![](./static/service_parallel_infra_seq.png)

Each service is deployed sequentially, but the infrastructures are deployed in parallel.

The deployment sequence is as follows:

Here, 1 and 2 will be deployed parallelly and 3 and 4 will be deployed parallelly i.e different services are deployed parallelly in the same infrastructure. 1 and 3 will deployed sequentially and 2 and 4 will be deployed sequentially i.e same service will be deployed in different infrastructure sequentially.

![](./static/seq_service_parallel_infra.png)

### Multi Service in Parallel and Multi Infrastructure in Parallel

- **Services**: S1, S2
- **Environment**: E1
- **Infrastructure**: I1, I2

Both services and infrastructure are deployed in parallel.

The deployment sequence is as follows:
**`S1 I1, S2 I1, S1 I2, S2 I2`**

![](./static/service_parallel_infra_parallel.png)

### Multi service in parallel and multi environment and multi Infrastructure in sequential 

> 

Reviewer 2: Add the sample title here

- **Services**: S1, S2
- **Environment_1_Infrastructure_1**: I1, I2
- **Environment_2_Infrastructure_2**: K82, K81

![](./static/multi_service_parallel_multi_env_infra_seq.png)

> Services are deployed in parallel, but the environments and infrastructures are deployed sequentially.

Reviewer 2: This line needs to be rephrased

Reviewer 2: Services are deployed in parallel sequentially in the infrastructures
